ncm-openstack/src/main/pan/components/openstack/schema.pan

```diff
@@ -0,0 +1,666 @@
+# ${license-info}
+# ${developer-info}
+# ${author-info}
+
+
+declaration template components/openstack/schema;
+
+include 'quattor/schema';
+include 'pan/types';
+
+
+function is_consistent_glance_store = {
+    # check store values
+    foreach (key; value; SELF) {
+        foreach (key2; value2; SELF) {
+            if (SELF[key] == SELF[key2] && key != key2) {
+                error(format("entry: %s appears several times within glance_store/stores list", value));
+            };
+        };
+    };
+    foreach (key; value; SELF) {
+        if (! match('^(file|http|swift|rbd|sheepdog|cinder|vmware)$', SELF[key])) {
+            error(format("entry: %s is not a valid Glance store value", value));
+        };
+    };
+    true;
+};
+
+function is_consistent_neutron_driver = {
+    foreach (key; value; SELF) {
+        if (! match('^(local|flat|vlan|gre|vxlan|geneve)$', SELF[key])) {
+            error(format("entry: %s is not a valid Neutron type driver", value));
+        };
+    };
+    true;
+};
+
+function is_consistent_neutron_extension_driver = {
+    foreach (key; value; SELF) {
+        if (! match('^(qos|port_security)$', SELF[key])) {
+            error(format("entry: %s is not a valid Neutron extension driver", value));
+        };
+    };
+    true;
+};
+
+@documentation {
+    OpenStack common domains section
+}
+type openstack_domains_common = extensible {
+    @{Domain name containing project}
+    'project_domain_name' : string = 'Default'
+    @{Project name to scope to}
+    'project_name' : string = 'service'
+    @{The type of authentication credential to create.
+    Required if no context is passed to the credential factory}
+    'auth_type' : string = 'password' with match (SELF, '^(password|token|keystone_token|keystone_password)$')
+    @{Users domain name}
+    'user_domain_name' : string = 'Default'
+    @{Keystone authentication URL http(s)://host:port}
+    'auth_url' : type_absoluteURI
+    @{OpenStack service username}
+    'username' : string
+    @{OpenStack service user password}
+    'password' : string
+};
+
+@documentation {
+    The revoke configuration section
+}
+type openstack_keystone_revoke = extensible {
```

Comment:
i propose to move all service specific types to their own declaration templates.